### ISSUE TITLE
rc7.1.1 Invalid Authentication, was working on rc7 and pr1.5

### ISSUE THREAD

**Issue by Snuffy2** [gcobb321/icloud3_v3#221]
It turns out I never was actually running rc6 due to the issues you noted in #219

But yesterday I was running rc7 and was not getting Authentication errors. I had downgraded to rc6 (actually pr1.5) after the other errors in #219 and ran that without issues until ~6pm tonight. I then upgraded to rc7.1.1 and restarted HA and got the Authentication errors in the log. 

I'll see if I somehow tried to login too much and got blocked by apple somehow and if it starts working later tonight/tomorrow. Tomorrow, I'll also try reverting to rc6 again if I'm still not able to authenticate.

[icloud3-0.log](https://github.com/gcobb321/icloud3_v3/files/13048552/icloud3-0.log)

**Comment by gcobb321** [comment]
I beat on the reauthentication pretty hard doing some debugging Tuesday without any problems testing invalid p/w entry with wrong p/w and various confirmation screens. Never had an issue. 

Try going back into _Configure Settings > Enter/Request Verification Code_ screen and request a code. Enter and go from there. 

A lot of times, when I entered the wrong p/w, it asked for a new code. Maybe your ‘verification code neededz’ was a coincidence due to the reinstalls.

**Comment by Snuffy2** [comment]
Sorry, I forgot to mention that, brain fart. I did try going into the options and logging in again. I ensured name/pass was correct and still got the error. I tried to resend a new code and that also didn't work.

<img width="615" alt="2023-10-19_20-42-17 260" src="https://github.com/gcobb321/icloud3_v3/assets/6526076/9a3c3b0d-ab01-46fe-ae34-37905955d53e">

<img width="616" alt="2023-10-19_20-42-42 361" src="https://github.com/gcobb321/icloud3_v3/assets/6526076/c4edfc45-94c3-4e24-b3ee-24be225a075d">

**Comment by gcobb321** [comment]
Unknown error is a bug. There should be error messages in the HA and iCloud-0.log files. Devices & services > system > logs might also have some info. Send me anything you can. 

Țry _Event Log > Actions > Rest iCloud Interface _ which will delete cookies and tokens which will generate a new code. HA should then highlight the notification bell and take you to a code entry screen.

**Comment by Snuffy2** [comment]
Sorry, I don't see the option to Reset the iCloud Interface in the Event Log Actions. Unknown error details below.
<img width="758" alt="2023-10-19_21-16-23" src="https://github.com/gcobb321/icloud3_v3/assets/6526076/72fee7a6-1dc6-4028-ba25-cdf1464e3716">

<img width="743" alt="2023-10-19_21-16-54" src="https://github.com/gcobb321/icloud3_v3/assets/6526076/f8feea3a-0e03-4df3-92cf-5dd47f934262">

Unknown error details:
```
2023-10-19 19:43:17.569 ERROR (MainThread) [icloud3-cf] Error logging into iCloud service: Login Error (Invalid username/password)/609
2023-10-19 19:43:34.189 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 85, in security_filter_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 80, in ban_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 236, in auth_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/headers.py", line 31, in headers_middleware
    response = await handler(request)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 148, in handle
    result = await handler(request, **request.match_info)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/decorators.py", line 63, in with_admin
    return await func(self, request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 234, in post
    return await super().post(request, flow_id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 72, in wrapper
    result = await method(view, request, data, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 110, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 293, in async_configure
    result = await self._async_handle_step(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 394, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/icloud3/config_flow.py", line 2098, in async_step_reauth
    if self.username == Gb.PyiCloud.username and self.password == Gb.PyiCloud.password:
                        ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'username'
```

**Comment by gcobb321** [comment]
I think what was happening is several errors happen at once that were not being caught along with a sequence of steps I didn’t anticipate. 

The Gb.PyIcloud is the iCloud session logged into when iCloud3 starts. This was not done because of a missing username or p/w. When it was entered on the config screen a different iCloud session is set up in case the username or p/w was for another account. That generates a verification code request that was checking the startup session for its info, which was not done. I found that if you tried the username or p/w without logging in first, the Gb. variables were not set up and generated the error. I don’t know if that was the sequence you did but that has been fixed. 

Doing a login first will login and then take you to the code entry screen but the soon to be attached module should prevent that error. 

Unzip it into the _icloud3_ directory and restart HA.
